### DISCUSSION TITLE
LibraryImport and array of bool values

### DISCUSSION THREAD

**Discussion by smourier** [dotnet/runtime#101494] [Q&A] [answered]
Hi,

I understand that there's an obligation to specify `MarshalAs` when using bools like here:

        [LibraryImport("somelib")]
        public static partial int SomeApi([MarshalAs(UnmanagedType.Bool)] bool one);

That's fine and dandy.
But is there a simple way to declare an array of bools? like here which obviously won't compile since we never say how `bool` type must be marshalled::

        [LibraryImport("somelib")]
        public static partial int SomeApi(out int count, out bool[] some);

This, which is nice for supported array types (like `int`),  doesn't compile any better (again, we don't say how `bool` type is marshalled):

        [LibraryImport("somelib")]
        public static partial int SomeApi(out int count, [MarshalUsing(typeof(ArrayMarshaller<,>), CountElementName = nameof(count))] out bool[] some);

And this doesn't work either:

        public static partial int SomeStructApi(out int count, [MarshalAs(UnmanagedType.Bool)][MarshalUsing(typeof(ArrayMarshaller<,>), CountElementName = nameof(count))] out bool[] some);

So do we have to write a custom marshaller or is there a simpler way to pass the way we want to marshal the bools?

**Comment by jkoritzinsky** [comment, marked as answer]
You can use `[MarshalAs(UnmanagedType.LPArray, ArraySubType = UnmanagedType.Bool)]` for arrays of bools.

**Reply by smourier** [reply]
Thanks! I know this pattern very well with runtime marshalling - well, not for bool since it wasn't needed - but I wouldn't have thought about it now when it's disabled.

There are now so many combination of what we can do between MarshalAs which has many limits), MarshalUsing, already written custom marshallers for arrays (fixed or not), collection, for com objects, etc., some existing pattersn still work, some don't... IMHO, it could be valuable to just present a list of like 50 frequent use cases / samples in official documentation, just like this one.
